### DISCUSSION TITLE
Proposal: Json Schema using PHP Type Declarations

### DISCUSSION THREAD

**Discussion by bruceoutdoors** [swaggest/php-json-schema#146] [Ideas]
Hi @vearutop,

I figured this library could leverage PHP type declarations to build the JSON schema, and use attributes for everything else; this would largely replace what `setUpProperties` is doing.

Here's a basic example:

```php
class UserOptions extends ClassStructure
{
    public int $id;
    public bool $auto_login;
    public string $group_name;
    public string $created_at;
}
```

Using PHP's Reflection API we can infer the following JSON schema:

```json
{
  "properties": {
    "id": {
      "type": "integer"
    },
    "auto_login": {
      "type": "boolean"
    },
    "group_name": {
      "type": "string"
    },
    "created_at": {
      "type": "string"
    }
  },
  "type": "object"
}
```

To add other constraints, use attributes:

```php
#[JSchema(['required'  => ['id', 'group_name']])]
class UserOptions extends AutoClassStructure
{
    public int $id;
    public bool $auto_login;
    public string $group_name;
    #[JSchema(['format' => 'date-time'])]
    public string $created_at;
}
```

And this will result in the following schema:

```json
{
  "required": [
    "id",
    "group_name"
  ],
  "properties": {
    "id": {
      "type": "integer"
    },
    "auto_login": {
      "type": "boolean"
    },
    "group_name": {
      "type": "string"
    },
    "created_at": {
      "type": "string",
      "format": "date-time"
    }
  },
  "type": "object"
}
```

Of course, implementation could be rather complex when you factor inheritance, union types, arrays, and enums. I have a working POC that extends `ClassStructure` in [this GitHub Gist](https://gist.github.com/bruceoutdoors/b7c207788cb47cf1f0271625128ad289). It utilizes features in PHP 8.1, so such a feature _would_ break compatibility with all older PHP versions.

I will be open to contribute this feature, but do let me know your thoughts on this!

**Comment by vearutop** [comment]
This looks like a great modernization idea and an improvement to API usability. 

Do you think it would be possible to guard such functionality with some feature checks (or options) to keep backwards compatibility?

(Sorry, I wasn't really following recent PHP development and lost some fluency.) 😅
